tempesta_fw/http.h

```diff
@@ -285,73 +285,53 @@ enum {
-	TFW_HTTP_B_CONN_EXTRA,
+	TFW_HTTP_CONN_EXTRA,
 	/* Chunked transfer encoding. */
-	TFW_HTTP_B_CHUNKED,
+	TFW_HTTP_CHUNKED,
 	/* Singular header presents more than once. */
-	TFW_HTTP_B_FIELD_DUPENTRY,
+	TFW_HTTP_FIELD_DUPENTRY,
 
 	/* Request flags. */
 	TFW_HTTP_FLAGS_REQ,
 	/* Sticky cookie is found and verified. */
-	TFW_HTTP_B_HAS_STICKY	= TFW_HTTP_FLAGS_REQ,
+	TFW_HTTP_HAS_STICKY	= TFW_HTTP_FLAGS_REQ,
 	/* URI has form http://authority/path, not just /path */
-	TFW_HTTP_B_URI_FULL,
+	TFW_HTTP_URI_FULL,
 	/* Request is non-idempotent. */
-	TFW_HTTP_B_NON_IDEMP,
+	TFW_HTTP_NON_IDEMP,
 	/* Request was sent by attacker. */
-	TFW_HTTP_B_SUSPECTED,
+	TFW_HTTP_SUSPECTED,
 	/* Request stated 'Accept: text/html' header */
-	TFW_HTTP_B_ACCEPT_HTML,
+	TFW_HTTP_ACCEPT_HTML,
 	/* Request is created by HTTP health monitor. */
-	TFW_HTTP_B_HMONITOR,
+	TFW_HTTP_HMONITOR,
 	/* Request from whitelist: skip frang and sticky modules processing. */
-	TFW_HTTP_B_WHITELIST,
+	TFW_HTTP_WHITELIST,
 	/* Client was disconnected, drop the request. */
-	TFW_HTTP_B_REQ_DROP,
+	TFW_HTTP_REQ_DROP,
 
 	/* Response flags */
 	TFW_HTTP_FLAGS_RESP,
 	/* Response has no body. */
-	TFW_HTTP_B_VOID_BODY	= TFW_HTTP_FLAGS_RESP,
+	TFW_HTTP_VOID_BODY	= TFW_HTTP_FLAGS_RESP,
 	/* Response has header 'Date:'. */
-	TFW_HTTP_B_HDR_DATE,
+	TFW_HTTP_HDR_DATE,
 	/* Response has header 'Last-Modified:'. */
-	TFW_HTTP_B_HDR_LMODIFIED,
+	TFW_HTTP_HDR_LMODIFIED,
 	/* Response is stale, but pass with a warning. */
-	TFW_HTTP_B_RESP_STALE,
+	TFW_HTTP_RESP_STALE,
 	/* Response is fully processed and ready to be forwarded to the client. */
-	TFW_HTTP_B_RESP_READY,
+	TFW_HTTP_RESP_READY,
 
 	_TFW_HTTP_FLAGS_NUM
 };
 
-#define TFW_HTTP_F_CONN_CLOSE		(1U << TFW_HTTP_B_CONN_CLOSE)
-#define TFW_HTTP_F_CONN_KA		(1U << TFW_HTTP_B_CONN_KA)
-#define __TFW_HTTP_MSG_M_CONN_MASK	\
-	(TFW_HTTP_F_CONN_KA | TFW_HTTP_F_CONN_CLOSE)
-#define TFW_HTTP_F_CONN_EXTRA		(1U << TFW_HTTP_B_CONN_EXTRA)
-#define TFW_HTTP_F_CHUNKED		(1U << TFW_HTTP_B_CHUNKED)
-#define TFW_HTTP_F_FIELD_DUPENTRY	(1U << TFW_HTTP_B_FIELD_DUPENTRY)
-
-#define TFW_HTTP_F_HAS_STICKY		(1U << TFW_HTTP_B_HAS_STICKY)
-#define TFW_HTTP_F_URI_FULL		(1U << TFW_HTTP_B_URI_FULL)
-#define TFW_HTTP_F_NON_IDEMP		(1U << TFW_HTTP_B_NON_IDEMP)
-#define TFW_HTTP_F_SUSPECTED		(1U << TFW_HTTP_B_SUSPECTED)
-#define TFW_HTTP_F_ACCEPT_HTML		(1U << TFW_HTTP_B_ACCEPT_HTML)
-#define TFW_HTTP_F_HMONITOR		(1U << TFW_HTTP_B_HMONITOR)
-#define TFW_HTTP_F_WHITELIST		(1U << TFW_HTTP_B_WHITELIST)
-#define TFW_HTTP_F_REQ_DROP		(1U << TFW_HTTP_B_REQ_DROP)
-
-#define TFW_HTTP_F_VOID_BODY		(1U << TFW_HTTP_B_VOID_BODY)
-#define TFW_HTTP_F_HDR_DATE		(1U << TFW_HTTP_B_HDR_DATE)
-#define TFW_HTTP_F_HDR_LMODIFIED	(1U << TFW_HTTP_B_HDR_LMODIFIED)
-#define TFW_HTTP_F_RESP_STALE		(1U << TFW_HTTP_B_RESP_STALE)
-#define TFW_HTTP_F_RESP_READY		(1U << TFW_HTTP_B_RESP_READY)
+#define __TFW_HTTP_MSG_M_CONN_MASK					\
+	(BIT(TFW_HTTP_CONN_CLOSE) | BIT(TFW_HTTP_CONN_KA))
```

Comment:
On one hand, these flags are mutually exclusive and they are handled and used in the code as mutually exclusive flags, i.e. if one is set, then do this, else do that.
On the other hand, there's an implicit understanding in the code that the `TFW_HTTP_CONN_CLOSE` flag takes precedence over the `TFW_HTTP_CONN_KA` flags.

I think a thorough comment is need to clearly state the rules of engagement with these two flags. Perhaps, in the enum where these flags are defined. Do they have to be maintained as mutually exclusive? Is it possible that both are set after a message is parsed? Does one takes precedence over the other?

Reply:
@keshonok many thanks for review!

Hm, yes, comment for `CONN_CLOSE` and `CONN_KA` is required. This flags actually has two different meanings in the same time and it always was in that way. 

`CONN_CLOSE` flag has two meanings:
- `Connection:` header contains `close` term;
- Connection is to be closed after this message. This flag is also set when there is no `Connection: close` term, i.e. for HTTP/1.0 connections,  or when error/attack has spotted and client connection is to be closed with a proper error response.

`CONN_KA` flag has only one meaning:
-   `Connection:` header contains `keep-alive` term.
Unlike to `CONN_CLOSE` there is no 'connection is in keep-alive mode'. It's implicitly so for HTTP/1.1 connections, so no special flag is required.

This two meanings are used for two aims:
- to init `SS_FLAG` when message is forwarded: to tell lower kernel levels to close the connection after transmission. We just need to know whether the `CONN_CLOSE` flag is set or not.
- to adjust `Connection:` header before forwarding. We would like to know about current content of the `Connection:` header, and these flags are used to take decision whether rewriting of the header is required or not.

Hard to say should we require mutual exclusivity for the flags or not. On the one hand there only two cases: close connection or keep it. No other variants. On the other hand, if the message originally had `Connection: keep-alive` header, but Tempesta decided to close the connection after it,  `keep-alive` token is still here and we have to modify the connection header. E.g.:
- just parsed request has `Connection: keep-alive` header;
- error happened while allocating sibling request message;
- we set `Conn_Stop` flag to the connection to stop further parsing operations and keep connection open to process all pending requests; 
- set `CONN_CLOSE` flag to the current request to close the connection once corresponding response is forwarded.

So I've updated the code in such a way, that mutual exclusivity is not required now. Precedence of `CONN_CLOSE` over `CONN_KA` flag is less strict but still effective condition.